File: DATA-HUB/Services/TableauVisualizationInfoExtractor.cs
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Xml.Linq;
using Newtonsoft.Json.Linq;
using OfficeOpenXml;

public class TableauVisualizationInfoExtractor
{
    private const string TableauCloudUrl = "https://10ay.online.tableau.com";
    private const string ApiVersion = "3.24"; 
    private const string TokenName = "PLACEHOLDER"; // Generate from Tableau account
    private const string TokenSecret = "PLACEHOLDER"; // Generate from Tableau account
    private const string SiteId = "tableauuser6-41ea9b971e";

    public async Task<List<VisualizationInfo>> ExtractVisualizationInfo(string filePath, string workbookName) {
        // Set up
        List<VisualizationInfo> visualizationInfoList = new List<VisualizationInfo>();
        var xmlDoc = LoadXmlDocument(filePath);
        var worksheets = GetWorksheetElements(xmlDoc);
        var authToken = await Authenticate();
        var url = $"{TableauCloudUrl}/api/metadata/graphql";
        using var client = new HttpClient();

        foreach (var worksheet in worksheets) {
            var worksheetName = GetWorksheetName(worksheet);
            var visualizationTitle = GetVisualizationTitle(worksheet);
            var visualizationType = GetVisualizationType(worksheet);
            string tablesUsed;
            string columnsUsed;
            if (visualizationType == "No Visualization") {
                tablesUsed = "No tables used";
                columnsUsed = "No columns used";
            } else {
                var metadata = await QueryMetadataApi(authToken, workbookName, worksheetName, url, client);
                JObject jsonMetadata = JObject.Parse(metadata);
                JArray tablesAsJson = (JArray)jsonMetadata["data"]["workbooks"][0]["sheets"][0]["upstreamTables"];
                JArray columnsAsJson = (JArray)jsonMetadata["data"]["workbooks"][0]["sheets"][0]["upstreamColumns"];
                tablesUsed = string.Join(", ", tablesAsJson.Select(t => t["name"]));
                columnsUsed = string.Join(", ", columnsAsJson.Select(c => c["name"]));
            }
            visualizationInfoList.Add(new VisualizationInfo {
                WorksheetName = worksheetName,
                VisualizationTitle = visualizationTitle,
                VisualizationType = visualizationType,
                TablesUsed = tablesUsed,
                ColumnsUsed = columnsUsed
            });
        }
        return visualizationInfoList;
    }

    public async Task<string> QueryMetadataApi(string authToken, string workbookName, string worksheetName, string url, HttpClient client) {
        var query = $@"
        {{
            workbooks (filter: {{ name: ""{workbookName}"" }}) {{
                sheets (filter: {{name: ""{worksheetName}"" }}) {{
                    upstreamTables {{
                        name
                    }}
                    upstreamColumns {{
                        name
                    }}
                }}
            }}
        }}";
        client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", authToken);

        var payload = new { query };
        var response = await client.PostAsync(url,
            new StringContent(JObject.FromObject(payload).ToString(), Encoding.UTF8, "application/json"));
        response.EnsureSuccessStatusCode();
        return await response.Content.ReadAsStringAsync();
    }

    public void SaveVisualizationInfoToJSONAndExcel(List<VisualizationInfo> visualizationInfoList, string jsonFilePath, string excelFilePath) {
        SaveToJSON(visualizationInfoList, jsonFilePath);
        SaveToExcel(visualizationInfoList, excelFilePath);
    }

    private XDocument LoadXmlDocument(string filePath) {
        return XDocument.Load(filePath);
    }

    private IEnumerable<XElement> GetWorksheetElements(XDocument xmlDoc) {
        return xmlDoc.Descendants("worksheet");
    }

    private string GetWorksheetName(XElement worksheet) {
        return worksheet.Attribute("name").Value;
    }
    
    private string GetVisualizationTitle(XElement worksheet) {
        var title = worksheet.Descendants("run").FirstOrDefault();
        if (title == null) {
            return "No Title";
        }
        return worksheet.Descendants("run").FirstOrDefault().Value;
    }

    private string GetVisualizationType(XElement worksheet) {
        var vizExists = worksheet.Descendants("datasource").FirstOrDefault();
        if (vizExists == null) {
            return "No Visualization";
        } else {
            return worksheet.Descendants("mark").FirstOrDefault().Attribute("class").Value;
        }
    }

    private void SaveToExcel(List<VisualizationInfo> visualizationInfoList, string excelFilePath) {
        using var package = new ExcelPackage();
        var worksheet = package.Workbook.Worksheets.Add("VisualizationInfo");
        AddExcelHeaders(worksheet);
        PopulateExcelData(worksheet, visualizationInfoList);
        worksheet.Cells[worksheet.Dimension.Address].AutoFitColumns();
        package.SaveAs(excelFilePath);
    }

    private void AddExcelHeaders(ExcelWorksheet worksheet) {
        worksheet.Cells[1, 1].Value = "Worksheet Name";
        worksheet.Cells[1, 2].Value = "Visualization Title";
        worksheet.Cells[1, 3].Value = "Visualization Type";
        worksheet.Cells[1, 4].Value = "Tables Used";
        worksheet.Cells[1, 5].Value = "Columns Used";
        worksheet.Row(1).Style.Font.Bold = true;
    }

    private void PopulateExcelData(ExcelWorksheet worksheet, List<VisualizationInfo> visualizationInfoList) {
        for (int i = 0; i < visualizationInfoList.Count; i++) {
            worksheet.Cells[i + 2, 1].Value = visualizationInfoList[i].WorksheetName;
            worksheet.Cells[i + 2, 2].Value = visualizationInfoList[i].VisualizationTitle;
            worksheet.Cells[i + 2, 3].Value = visualizationInfoList[i].VisualizationType;
            worksheet.Cells[i + 2, 4].Value = string.Join(", ", visualizationInfoList[i].TablesUsed);
            worksheet.Cells[i + 2, 5].Value = string.Join(", ", visualizationInfoList[i].ColumnsUsed);
        }
    }

    private void SaveToJSON(List<VisualizationInfo> visualizationInfoList, string jsonFilePath) {
        string json = JsonSerializer.Serialize(visualizationInfoList, new JsonSerializerOptions { WriteIndented = true });
        File.WriteAllText(jsonFilePath, json);
    }

    private async Task<string> Authenticate() {
        using var client = new HttpClient();
        var url = $"{TableauCloudUrl}/api/{ApiVersion}/auth/signin";
        var payload = new
        {
            credentials = new
            {
                personalAccessTokenName = TokenName,
                personalAccessTokenSecret = TokenSecret,
                site = new { contentUrl = SiteId }
            }
        };

        var request = new HttpRequestMessage(HttpMethod.Post, url) {
            Content = new StringContent(JObject.FromObject(payload).ToString(), Encoding.UTF8, "application/json")
        };
        request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

        var response = await client.SendAsync(request);
        response.EnsureSuccessStatusCode();

        var json = JObject.Parse(await response.Content.ReadAsStringAsync());
        var authToken = json["credentials"]["token"].ToString();

        return authToken;
    }

    public class VisualizationInfo {
        public string WorksheetName { get; set; }
        public string VisualizationTitle { get; set; }
        public string VisualizationType { get; set; }
        public string TablesUsed { get; set; }
        public string ColumnsUsed { get; set; }
    }
}

File: DATA-HUB/Services/tableau_visualizationInfo_extractor.py
import json
from openpyxl.worksheet.worksheet import Worksheet
from openpyxl.styles import Font
from tableau_api_lib import TableauServerConnection
from collections import namedtuple
from openpyxl import Workbook
import xml.etree.ElementTree as ET
import os
import time

class VisualizationInfo:
    def __init__(self, worksheet_name: str, viz_title: str, viz_type: str, tables_used: str, columns_used: str):
        self.worksheet_name = worksheet_name
        self.viz_title = viz_title
        self.viz_type = viz_type
        self.tables_used = tables_used
        self.column_used = columns_used

def get_workbook_name(file_path: str):
    file_name = os.path.basename(file_path)
    workbook_name, _ = os.path.splitext(file_name)
    return workbook_name

def get_viz_title(worksheet: ET.Element):
    title = worksheet.find(".//run")
    if title is None:
        return "No Title"
    return title.text

def get_viz_type(worksheet: ET.Element):
    vizExists = worksheet.find(".//datasource")
    if vizExists is None:
        return "No Visualization"
    return worksheet.find(".//mark").get("class")

def get_tables_and_columns(workbook_name: str, worksheet_name: str, connection: TableauServerConnection):
    query = f"""
    {{
        workbooks (filter: {{ name: "{workbook_name}"}}) {{
            sheets (filter: {{ name: "{worksheet_name}"}}) {{
                upstreamTables {{
                    name
                }}
                upstreamColumns {{
                    name
                }}
            }}
        }}
    }}
    """
    time.sleep(10)
    response = connection.metadata_graphql_query(query=query)
    response_json = response.json()

    tables_as_json = response_json["data"]["workbooks"][0]["sheets"][0]["upstreamTables"]
    columns_as_json = response_json["data"]["workbooks"][0]["sheets"][0]["upstreamColumns"]
    tablesUsed  = ", ".join(list(dict.values())[0] for dict in tables_as_json)
    columnsUsed = ", ".join(list(dict.values())[0] for dict in columns_as_json)
    return (tablesUsed, columnsUsed)

def save_to_excel(visualization_info_list: list, excel_file_path: str):
    workbook = Workbook()
    worksheet = workbook.active
    worksheet.title = "VisualizationInfo"
    add_excel_headers(worksheet)
    populate_excel_data(worksheet, visualization_info_list)
    for column_cells in worksheet.columns:
        max_length = max(len(str(cell.value)) if cell.value else 0 for cell in column_cells)
        worksheet.column_dimensions[column_cells[0].column_letter].width = max_length + 2
    workbook.save(excel_file_path)

def add_excel_headers(worksheet: Worksheet):
    headers = ["Worksheet Name", "Visualization Title", "Visualization Type", "Tables Used", "Columns Used"]
    worksheet.append(headers)
    for cell in worksheet[1]:
        cell.font = Font(bold=True)

def populate_excel_data(worksheet: Worksheet, visualization_info_list: list):
    for i, vis_info in enumerate(visualization_info_list, start=2):
        worksheet.cell(row=i, column=1, value=vis_info.WorksheetName)
        worksheet.cell(row=i, column=2, value=vis_info.VisualizationTitle)
        worksheet.cell(row=i, column=3, value=vis_info.VisualizationType)
        worksheet.cell(row=i, column=4, value=vis_info.TablesUsed)
        worksheet.cell(row=i, column=5, value=vis_info.ColumnsUsed)

def save_to_json(visualization_info_list: list, json_file_path: str):
    with open(json_file_path, "w", encoding="utf-8") as json_file:
        json.dump([info._asdict() for info in visualization_info_list], json_file, indent=4)

def extract_viz_info(workbook_file_path: str, packaged_workbook_file_path: str):
    config = {
    "tableau_prod": {
        "server": "https://10ay.online.tableau.com",
        "api_version": "3.22",
        "personal_access_token_name": "PLACEHOLDER",
        "personal_access_token_secret": "PLACEHOLDER",
        "site_name": "arjun-at-quadrant",
        "site_url": "arjun-at-quadrant"
        }
    }
    conn = TableauServerConnection(config_json=config, env="tableau_prod")
    conn.sign_in()
    project_name = "Demo"
    project_description = "This project contains workbooks for metadata extraction."
    project_info = conn.create_project(project_name=project_name, project_description=project_description)
    project_info_json = project_info.json()
    project_id = project_info_json["project"]["id"]
    workbook_name = get_workbook_name(packaged_workbook_file_path)
    conn.publish_workbook(workbook_file_path=packaged_workbook_file_path, workbook_name=workbook_name, project_id=project_id, connection_username="PLACEHOLDER", connection_password="PLACEHOLDER")
    workbook_xml_doc = ET.parse(workbook_file_path)
    workbook_worksheets = workbook_xml_doc.findall(".//worksheet")
    viz_info_list = []
    for worksheet in workbook_worksheets:
        worksheet_name = worksheet.get("name")
        viz_title = get_viz_title(worksheet)
        viz_type = get_viz_type(worksheet)
        tables_used = None
        columns_used = None
        if (viz_type == "No Visualization"):
            tables_used = "No Tables Used"
            columns_used = "No Columns Used"
        else:
            (tables_used, columns_used) = get_tables_and_columns(workbook_name=workbook_name, worksheet_name=worksheet_name, connection=conn)
        viz_info = VisualizationInfo(worksheet_name, viz_title, viz_type, tables_used, columns_used)
        viz_info_list.append(viz_info)

def save_viz_info_to_json_and_excel(viz_info: str, json_file_path: str, excel_file_path: str):
    save_to_json(viz_info, json_file_path)
    save_to_excel(viz_info, excel_file_path)